sys/frac/frac.c

```diff
@@ -0,0 +1,157 @@
+/**
+ * Copyright (C) 2018 Eistec AB
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ *
+ * @ingroup sys_util
+ * @{
+ * @file
+ * @brief    Integer fraction function implementations
+ *
+ * @author   Joakim Nohlgård <joakim.nohlgard@eistec.se>
+ *
+ * @}
+ */
+
+#include <stdint.h>
+#include <stdio.h>
+#include "assert.h"
+#include "frac.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+/**
+ * @brief   compute greatest common divisor of @p u and @p v
+ *
+ * @param[in]   u    first operand
+ * @param[in]   v    second operand
+ *
+ * @return  Greatest common divisor of @p u and @p v
+ */
+static uint32_t gcd32(uint32_t u, uint32_t v)
+{
+    /* Source: https://en.wikipedia.org/wiki/Binary_GCD_algorithm#Iterative_version_in_C */
+    unsigned shift;
+
+    /* GCD(0,v) == v; GCD(u,0) == u, GCD(0,0) == 0 */
+    if (u == 0) {
+        return v;
+    }
+    if (v == 0) {
+        return u;
+    }
+
+    /* Let shift := log2 K, where K is the greatest power of 2
+     * dividing both u and v. */
+    for (shift = 0; ((u | v) & 1) == 0; ++shift) {
+        u >>= 1;
+        v >>= 1;
+    }
+
+    /* remove all factors of 2 in u */
+    while ((u & 1) == 0) {
+        u >>= 1;
+    }
+
+    /* From here on, u is always odd. */
+    do {
+        /* remove all factors of 2 in v -- they are not common */
+        /*   note: v is not zero, so while will terminate */
+        while ((v & 1) == 0) {
+            v >>= 1;
+        }
+
+        /* Now u and v are both odd. Swap if necessary so u <= v,
+         * then set v = v - u (which is even). */
+        if (u > v) {
+            /* Swap u and v */
+            unsigned int t = v;
```

Comment:
Possible loss of precision, as `v` is `uint32_t` and `unsigned int` can be 16 bits wide

Reply:
yes, fixed.